### DISCUSSION TITLE
generate dfu-bin bootloaer or a combined binary file

### DISCUSSION THREAD

**Discussion by vercingetor-ix** [blackmagic-debug/blackmagic#1784] [Q&A]
Hi,
when i flashed my last bmp some years ago i used the "make" command to create my binaries.
The result was two files, a _***.bin**_ and a ***_dfu.bin** file. The latter was the bootloader, I loaded with the serail adapter to flash the bmp firmware in next step.
When I now create the binaries with meson, there is an empty "blackmagic_bluepill_bootloader.elf.p" directory below the build directory, but no bootloader binary, only the firmware binary.
I searched the documentation, but found no information how to create now the dfu-bootloader.
I am thankful for any hint, how to proceed.

**Comment by dragonmux** [comment]
👋🏼 It's not built by default as there's only a very narrow set of times when you actually need to build the bootloader (it changes very rarely and isn't part of the normal update process once put on a part the first time).

If you absolutely need to regenerate it, it's `ninja boot-elf` for the .elf and `ninja boot-bin` for the .bin from inside the build directory. There's likely a way to do this from `meson compile -C build`, but we weren't about to go diving the Meson docs to check. Hopefully that helps.

**Reply by vercingetor-ix** [reply]
Thanks for your fast answer, that a helped a lot!
Its necessary, as I will flash a bare new bluepill board, which has no bootloader.
I was not able to figure out, how to do that with meson/ninja, as both tools are new for me!

**Reply by dragonmux** [reply]
Understandable, we'll work with Perigoso on documenting the bootloader aspect properly. The very short version is that any time you see `-C build` in the documentation, you can in your head rewrite it to `cd build && <command>` with the `-C build` part chopped off for `<command>`.

Ninja has a way to list out the available targets in the form of `ninja -t targets` which gives a nice clean list of all the things the build system registers that you can build. Hope that helps too.
